### PR TITLE
code optimization to reduce memory copies and loops when seriesHashmap del

### PR DESCRIPTION
**Pull request by songjiayang**

### CODE REVIEW

File: tsdb/head.go [hunk 1]
```diff
@@ -1706,17 +1706,27 @@ func (m seriesHashmap) set(hash uint64, s *memSeries) {
 }
 
 func (m seriesHashmap) del(hash uint64, lset labels.Labels) {
-	var rem []*memSeries
-	for _, s := range m[hash] {
-		if !labels.Equal(s.lset, lset) {
-			rem = append(rem, s)
+	for index, s := range m[hash] {
+		if labels.Equal(s.lset, lset) {
```
Comment (nicolastakashi) [inline tsdb/head.go:1710]: wdyt?
early returns and also reduce cyclomatic complexity
```suggestion
		if !labels.Equal(s.lset, lset) {
		    continue
		}
```

**Comment by bboreham** [comment]
Thanks for your PR.   In the PR template text it says "Performance improvements would need a benchmark test to prove it."
Can you supply benchmark numbers please?

Note hash collisions (the case where the loop has more than 1 iteration) are very rare. 
Also this code will change if #13040 is merged.

File: tsdb/head.go (hunk 1 above)
Comment (songjiayang) [reply tsdb/head.go:1710]: thanks

**Comment by songjiayang** [comment]
> Thanks for your PR. In the PR template text it says "Performance improvements would need a benchmark test to prove it." Can you supply benchmark numbers please?
> 
> Note hash collisions (the case where the loop has more than 1 iteration) are very rare. Also this code will change if #13040 is merged.

Hi @bboreham  thanks for your review.

With this benchmark codes

```
func BenchmarkSeriesHashmapDelWithNew(b *testing.B) {
	hash := uint64(0)
	count := 3000

	for i := 0; i < b.N; i++ {
		m := generateSeriesHashmap(hash, count)
		m.del(hash, m[hash][10].lset)
		m.del(hash, m[hash][100].lset)
		m.del(hash, m[hash][1000].lset)
		m.del(hash, m[hash][2000].lset)
		m.del(hash, m[hash][2990].lset)
	}
}

func BenchmarkSeriesHashmapDelOld(b *testing.B) {
	hash := uint64(0)
	count := 3000

	for i := 0; i < b.N; i++ {
		m := generateSeriesHashmap(hash, count)
		m.oldDel(hash, m[hash][10].lset)
		m.oldDel(hash, m[hash][100].lset)
		m.oldDel(hash, m[hash][1000].lset)
		m.oldDel(hash, m[hash][2000].lset)
		m.oldDel(hash, m[hash][2990].lset)
	}
}

func generateSeriesHashmap(hash uint64, count int) seriesHashmap {
	h := seriesHashmap{}
	for i := 0; i < count; i++ {
		lset := []labels.Label{
			{
				Name:  fmt.Sprintf("label_name_%d", i),
				Value: fmt.Sprintf("label_name_%d", i),
			},
		}
		h.set(0, newMemSeries(lset, chunks.HeadSeriesRef(i+1), false))
	}
	return h
}
```

The result is:

```
go test -v  -bench="BenchmarkSeriesHashmapDel.*"  -run=^$ -count 5 -benchmem
goos: darwin
goarch: amd64
pkg: github.com/prometheus/prometheus/tsdb
cpu: Intel(R) Core(TM) i5-7360U CPU @ 2.30GHz
BenchmarkSeriesHashmapDelWithNew
BenchmarkSeriesHashmapDelWithNew-4   	      45	  24927126 ns/op	 1139866 B/op	   23506 allocs/op
BenchmarkSeriesHashmapDelWithNew-4   	      46	  30309610 ns/op	 1139845 B/op	   23506 allocs/op
BenchmarkSeriesHashmapDelWithNew-4   	      37	  35696554 ns/op	 1139859 B/op	   23506 allocs/op
BenchmarkSeriesHashmapDelWithNew-4   	      44	  32712634 ns/op	 1139831 B/op	   23506 allocs/op
BenchmarkSeriesHashmapDelWithNew-4   	      44	  30373490 ns/op	 1139850 B/op	   23506 allocs/op
BenchmarkSeriesHashmapDelOld
BenchmarkSeriesHashmapDelOld-4       	      45	  30474154 ns/op	 1576401 B/op	   23582 allocs/op
BenchmarkSeriesHashmapDelOld-4       	      36	  29201853 ns/op	 1576401 B/op	   23582 allocs/op
BenchmarkSeriesHashmapDelOld-4       	      38	  30110181 ns/op	 1576328 B/op	   23581 allocs/op
BenchmarkSeriesHashmapDelOld-4       	      42	  40354614 ns/op	 1576385 B/op	   23582 allocs/op
BenchmarkSeriesHashmapDelOld-4       	      46	  27857226 ns/op	 1576347 B/op	   23582 allocs/op
```

Through the result,   single hash slot has 3k series,  the new del method reduces memory about 25%.

**Comment by bboreham** [comment]
Thanks for the benchmark code.  I don't think it is realistic; hash collisions are a 1-in-a-billion thing.

**Comment by songjiayang** [comment]
Here I am not solving the problem of hash conflicts. I am improving and cleaning up the memory overhead and invalid loops caused by unnecessary series when performing head gc.

**Comment by bboreham** [comment]
Your benchmark fixes `hash` at zero.  `hash` should be computed from some set of labels to be realistic.

**Comment by songjiayang** [comment]
Sorry, I misunderstood. I regarded `seriesHashmap` as `[]*memSeries`.  Indeed, hash conflicts are rare, so  ignore the code please.

**Comment by songjiayang** [comment]
I'm very sorry for wasting your precious time. @bboreham

**Comment by bboreham** [comment]
Hey, don't feel bad!  It's totally fine to have ideas that don't work out.

**Pull request closed without merging**